fix(ngcc): handle compilation diagnostics

JoostK opened the pull request:
Previously, any diagnostics reported during the compilation of an
entry-point would not be shown to the user, but either be ignored or
cause a hard crash in case of a `FatalDiagnosticError`. This is
unfortunate, as such error instances contain information on which code
was responsible for producing the error, whereas only its error message
would not. Therefore, it was quite hard to determine where the error
originates from.

This commit introduces behavior to deal with error diagnostics in a more
graceful way. Such diagnostics will still cause the compilation to fail,
however the error message now contains formatted diagnostics.

Closes #31977
Resolves FW-1374

JoostK committed:
fix(ngcc): handle compilation diagnostics

Previously, any diagnostics reported during the compilation of an
entry-point would not be shown to the user, but either be ignored or
cause a hard crash in case of a `FatalDiagnosticError`. This is
unfortunate, as such error instances contain information on which code
was responsible for producing the error, whereas only its error message
would not. Therefore, it was quite hard to determine where the error
originates from.

This commit introduces behavior to deal with error diagnostics in a more
graceful way. Such diagnostics will still cause the compilation to fail,
however the error message now contains formatted diagnostics.

Closes #31977
Resolves FW-1374

petebacondarwin commented on packages/compiler-cli/ngcc/src/diagnostics.ts:
```diff
@@ -0,0 +1,24 @@
+  getCurrentDirectory: () => getFileSystem().pwd(),
+  getCanonicalFileName: fileName => fileName,
+  getNewLine: () => ts.sys.newLine,
+};
```
I think we could use the current `CompilerHost` instead of creating a new one here. And this would also have the benefit of matching any specialisation that we add to that host.

JoostK replied:
Excellent! I hadn't realized the `CompilerHost` interface is compatible with `FormatDiagnosticHost`.

petebacondarwin commented on packages/compiler-cli/ngcc/src/main.ts:
```diff
@@ -135,9 +138,18 @@ export function mainNgcc(
+          const result = transformer.transform(bundle);
+          if (result.success) {
+            if (result.diagnostics.length > 0) {
+              logger.warn(formatDiagnostics(result.diagnostics));
```
For example:

```suggestion
              logger.warn(ts.formatDiagnostics(result.diagnostics, bundle.src.host));
```

petebacondarwin reviewed (approved):
One suggestion to avoid having to make up our own `FormatDiagnosticHost`.

gkalpak reviewed (commented):
😍

JoostK commented:
Blocked on a legit Codefresh failure. Instead of fixing it right away, we'll first make sure the emulated Windows FS mode behaves identical to the native Windows FS mode as to catch such failures earlier in the future.

petebacondarwin commented:
Good man @JoostK !!

alxhub commented:
@JoostK sounds good! I'm gonna remove `merge-assistance` so this doesn't get merged until the codefresh failure is fixed.

JoostK committed:
test(ivy): normalize rooted paths to include a drive letter in Windows

The Angular compiler has an emulation system for various kinds of
filesystems and runs its testcases for all those filesystems. This
allows to verify that the compiler behaves correctly in all of the
supported platforms, without needing to run the tests on the actual
platforms.

Previously, the emulated Windows mode would normalize rooted paths to
always include a drive letter, whereas the native mode did not perform
this normalization. The consequence of this discrepancy was that running
the tests in native Windows was behaving differently compared to how
emulated Windows mode behaves, potentially resulting in test failures
in native Windows that would succeed for emulated Windows.

This commit adds logic to ensure that paths are normalized equally for
emulated Windows and native Windows mode, therefore resolving the
discrepancy.

JoostK commented:
Oh no! We don't currently have a Windows CI as Codefresh configuration was recently removed. I deliberately did not rebase my branch hoping that Codefresh would still run on the branch, but that doesn't appear to be the case :-(

@petebacondarwin I did add commit 3fdb4af which should resolve the discrepancy between emulated and native Windows mode. But unfortunately I can't test on native Windows now that Codefresh is gone...

petebacondarwin commented:
@gkalpak is sometimes able to run things on his Windows setup...

gkalpak commented on packages/compiler-cli/src/ngtsc/file_system/testing/src/mock_file_system_native.ts:
```diff
@@ -41,6 +45,15 @@ export class MockFileSystemNative extends MockFileSystem {
+    // (see `MockFileSystemWindows`) so that the behavior is identical between native Windows and
+    // emulated Windows mode.
+    if (isWindows) {
+      path = path.replace(/^[\/\\]/i, 'C:/') as T;
```
Nit: The `i` flag is redundant (as is escaping `/` inside `[...]` :grin:).

gkalpak commented:
As discussed offline:

I ran `yarn test-ivy-aot //packages/compiler-cli/ngcc/test/... //packages/compiler-cli/test/ngtsc` and all tests passed (but `ngcc/test:integration` is skipped due to its `no-ivy-aot` tag).

Running `yarn bazel test //packages/compiler-cli/ngcc/test/... //packages/compiler-cli/test/ngtsc` gives the following failure for `ngcc/test:integration`:

```
<<FileSystem: Native>> ngcc main() diagnostics should fail with formatted diagnostics when an error diagnostic is produced
  Message:
    Expected function to throw an exception with message 'Failed to compile entry-point fatal-error due to compilation errors:
    node_modules/fatal-error/index.js(5,17): error TS-992001: component is missing a template
    ', but it threw an exception with message 'Failed to compile entry-point fatal-error due to compilation errors:
    node_modules/fatal-error/index.js(5,17): error TS-992001: component is missing a template
    '.
```

The difference being that the actual error ends with `\r\n` instead of `\n`. So, changing [this assertion][1]:

```
.toThrowError(
    'Failed to compile entry-point fatal-error due to compilation errors:\n' +
    'node_modules/fatal-error/index.js(5,17): error TS-992001: component is missing a template\n');
```

...to something like...

```
.toThrowError(
    'Failed to compile entry-point fatal-error due to compilation errors:\n' +
    `node_modules/fatal-error/index.js(5,17): error TS-992001: component is missing a template${require('os').EOL}`);
```

...makes all tests pass.

##
Interestingly, hard-coding `\r\n` in the above assertion, also makes all tests pass on Windows, which probably means that the mock-fs is doing something wrong (e.g. not mocking the system's EOL property, so that `ts.formatDiagnostics()` always uses the native EOL instead of the mocked one).

##
I am not sure what is the best way to solve this (e.g. normalize newlines in [ts.formatDiagnostics()][2] or fix the assertions, etc.). but normalizing newlines to `\n` on the output of `ts.formatDiagnostics()` sounds reasonable to me (assuming it does not break something else :grin:).

[1]: https://github.com/angular/angular/blob/3ff6b7b90bb84df7567234929da5f2a4cb7fc23a/packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts#L439-L441
[2]: https://github.com/angular/angular/blob/3ff6b7b90bb84df7567234929da5f2a4cb7fc23a/packages/compiler-cli/ngcc/src/main.ts#L182

mhevery commented:
Is this still blocked?

mhevery commented:
[presubmit](http://test/OCL:266059678:BASE:266059696:1567053010198:d0536030)

gkalpak commented:
I just rerun the tests on Windows and they passed 👍

JoostK commented:
merge-assistance: job "legacy-unit-tests-saucelabs" is consistently timing out, unrelated to these changes.
